### ISSUE TITLE
Examples broken under pepper_canary

### ISSUE THREAD

**Issue by matthewturk** [google/pepper.js#4]
A fresh clone of this repo against pepper_canary doesn't build properly under examples, as it is looking for an `alarms_dev.cc` file that doesn't exist.

It looks like in canary the ppapi_cpp files have been moved around, but simply changing the `SOURCES` variable in `examples/ppapi_cpp/Makefile` to match the current contents of `$NACL_SDK_ROOT/src/ppapi_cpp/*.cc` fixes the problem and the examples build properly.

Once those have been fixed, the tests directory fails with missing `ppb_testing_dev.h` header, which I wasn't able to as easily correct.

**Comment by ncbray** [comment]
The fix should now be on master.  Does it work for you?

**Comment by matthewturk** [comment]
Not quite.  I'm not sure about the objects added inside the `$AT_LEAST_34` conditional -- I don't have any of those in my build of the pepper_34 SDK.  Should that be an `ifneq`?

**Comment by ncbray** [comment]
I did the patch against revision 243488.  pepper_canary is fairly volatile, so check if you have an older version:
./naclsdk info pepper_canary | grep revision

To get the latest:
./naclsdk update --force pepper_canary

Tell me if that fixes it.  Makefile is not my native language, so I do wish to be cautious on this issue.

**Comment by matthewturk** [comment]
Yup, that's the revision I'm on:

```
description: Chrome 34 bundle. Chrome revision: 243488. NaCl revision: 12558
```

To double check I wasn't compiling against an older version, I wiped the bundle and re-installed.  The error now seems to be `resource_array_dev.cc`, which I see in the current NaCl tree but not in the pepper_canary bundle.  I can now build against my checkout of chromium, though, so this is good for me.  I must have had a messed up path earlier.  Thank you!

**Comment by ncbray** [comment]
Uhg.  I just talked to the SDK folks and apparently "info" gives the revision for the _latest_ version of the SDK, not the local copy.  For posterity, ./naclsdk list -r gives the local revision.  I was actually on a slightly older version than I thought.

I'll do another patch for canary once 33 gets an official SDK release (build problems have held it up) so I can double check that, too.

Longer term, I'll need to find a way to import the file list directly from the SDK, but right now it just isn't set up for that.

**Comment by ncbray** [comment]
OK, the most recent version of pepper_canary should work now.  I also tested it on pepper_33.

**Comment by danielrh** [comment]
I downloaded pepper_34 and tried building the code against it, however the hello_world_gles test fails (the test works with pepper_33) because it tries to download an undefined filename. A relevant error could be the new Var interface

The error I'm getting on the console is 

```
Requested unknown interface: PPB_Var;1.2 hello_world_gles.js:21
View change: 640 x 480 hello_world_gles.js:114
Create instance: 261 ms common.js:203
Failed to load resource: the server responded with a status of 404 (Not Found) http://<myhostname>/pepper_34a/examples/hello_world_gles/data/undefined
Loaded 0
Loaded 0
Loaded 0
WebGL: INVALID_OPERATION: getUniformLocation: program not linked
WebGL: INVALID_OPERATION: getAttribLocation: program not linked
WebGL: INVALID_OPERATION: getUniformLocation: program not linked
...  
```

My build is here: http://graphics.stanford.edu/~danielrh/pepper_34a/examples/examples.html

Any ideas how to start debugging this?

**Comment by ncbray** [comment]
Based on that output, I suspect the new PPB_Var;1.2 interface needs to be supported.  (Although that error may have been due to interface probing and the error may be unrelated.)  Right now only up to 1.1 is implemented.  A bad Var interface pointer could cascade into requesting a bad URL, which appears to be what is happening.  Emcscripten is a little too nice about null pointer deferences and calling null function pointers.

If you're curious, a quick hack/diagnostic would be to duplicate this block:
https://github.com/google/pepper.js/blob/master/wrappers/base.js#L176
and give it the newer version number.  I believe the two new functions added to PPB_Var;1.2 should not be used by the example, so not implementing them should be OK.

I'll look into it myself, shortly, unless you want to take a swing at it yourself.

**Comment by danielrh** [comment]
Yes updating the interface version fixed the demo :-) Of course providing the new functions would be important to actually advertising that interface, but I'm glad it's something less subtle than I was worried... unfortunately pepper_35 errors out with a totally baffling problem:

```
  LINK emscripten/Release/geturl.js
Stack: Error
    at assertTrue (eval at globalEval (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:105:8), <anonymous>:59:26)
    at addTypeInternal (eval at globalEval (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:105:8), <anonymous>:953:7)
    at addType (eval at globalEval (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:105:8), <anonymous>:981:5)
    at typevestigator (eval at globalEval (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:105:8), <anonymous>:995:19)
    at analyzer (eval at globalEval (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:105:8), <anonymous>:1688:3)
    at JSify (eval at globalEval (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:105:8), <anonymous>:123:13)
    at runPhase (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:270:5)
    at compile (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:286:5)
    at Object.<anonymous> (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:296:5)
    at Module._compile (module.js:456:26)

undefined:60
    throw msg;
          ^
Assertion failed: 16 x i8
Stack: Error
    at assertTrue (eval at globalEval (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:105:8), <anonymous>:59:26)
    at addTypeInternal (eval at globalEval (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:105:8), <anonymous>:953:7)
    at addType (eval at globalEval (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:105:8), <anonymous>:981:5)
    at typevestigator (eval at globalEval (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:105:8), <anonymous>:995:19)
    at analyzer (eval at globalEval (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:105:8), <anonymous>:1688:3)
    at JSify (eval at globalEval (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:105:8), <anonymous>:123:13)
    at runPhase (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:270:5)
    at compile (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:286:5)
    at Object.<anonymous> (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:296:5)
    at Module._compile (module.js:456:26)

undefined:60
    throw msg;
          ^
Assertion failed: 16 x i8
Stack: Error
    at assertTrue (eval at globalEval (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:105:8), <anonymous>:59:26)
    at addTypeInternal (eval at globalEval (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:105:8), <anonymous>:953:7)
    at addType (eval at globalEval (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:105:8), <anonymous>:981:5)
    at typevestigator (eval at globalEval (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:105:8), <anonymous>:995:19)
    at analyzer (eval at globalEval (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:105:8), <anonymous>:1688:3)
    at JSify (eval at globalEval (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:105:8), <anonymous>:123:13)
    at runPhase (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:270:5)
    at compile (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:286:5)
    at Object.<anonymous> (/home/daniel/dev/pepper_35/emscripten/src/compiler.js:296:5)
    at Module._compile (module.js:456:26)

undefined:60
    throw msg;
          ^
Assertion failed: 4 x i32
Traceback (most recent call last):
  File "/home/daniel/dev/pepper_35/emscripten/emscripten.py", line 832, in <module>
    _main(environ=os.environ)
  File "/home/daniel/dev/pepper_35/emscripten/emscripten.py", line 820, in _main
    temp_files.run_and_clean(lambda: main(
  File "/home/daniel/dev/pepper_35/emscripten/tools/tempfiles.py", line 38, in run_and_clean
    return func()
  File "/home/daniel/dev/pepper_35/emscripten/emscripten.py", line 828, in <lambda>
    DEBUG_CACHE=DEBUG_CACHE,
  File "/home/daniel/dev/pepper_35/emscripten/emscripten.py", line 714, in main
    jcache=jcache, temp_files=temp_files, DEBUG=DEBUG, DEBUG_CACHE=DEBUG_CACHE)
  File "/home/daniel/dev/pepper_35/emscripten/emscripten.py", line 284, in emscript
    assert len(output) == 2, 'Did not receive forwarded data in an output - process failed? We only got: ' + output[0][-3000:]
AssertionError: Did not receive forwarded data in an output - process failed? We only got: Assertion failed: 16 x i8

Traceback (most recent call last):
  File "/home/daniel/dev/pepper_35/emscripten/emcc", line 1556, in <module>
    final = shared.Building.emscripten(final, append_ext=False, extra_args=extra_args)
  File "/home/daniel/dev/pepper_35/emscripten/tools/shared.py", line 1214, in emscripten
    assert os.path.exists(filename + '.o.js') and len(open(filename + '.o.js', 'r').read()) > 0, 'Emscripten failed to generate .js: ' + str(compiler_output)
AssertionError: Emscripten failed to generate .js: 
make[1]: *** [emscripten/Release/geturl.js] Error 1
make[1]: Leaving directory `/home/daniel/dev/pepper_35/examples/geturl'
make: *** [geturl_TARGET] Error 2
```

updating to any newer emscripten versions using (which I'd love to do to get simd support for my project) the default nacl clang setup results in the following error in the build tools

INFO     root: (Emscripten: Running sanity checks)
CRITICAL root: Cannot find /home/daniel/Downloads/nacl_sdk/nacl_sdk/pepper_35/toolchain/linux_pnacl/bin/pnacl-lli, check the paths in ~/.emscripten

I've not heard of the lli program and how it relates to the toolchain

Is it possible to build all the examples with pepper_35, maybe using a custom emscripten/clang setup? It would help switch back and forth between nacl since pepper_35 has the improved debugging facilities...

**Comment by ncbray** [comment]
Yeah, the LLI thing is kind of silly.  Emscripten accepted a patch that uses the LLVM interpreter to figure out the size of a structure, or something along those line.  PNaCl has been shipping a stripped down LLVM toolchain that did not include LLI.  This has prevented me from rolling the Emscripten version.  pepper_37 should finally have a toolchain that includes LLI.  Until then you can just set the LLVM environment variable to point to the bin directory of a full version of LLVM 3.2 (?) and roll Emscripten forward.  This has been bugging me for a while, but the true fix has not been easy.  :/

The errors you post look a lot like a SIMD support issue.  Keep me posted on if rolling Emscripten forward solves them.

In the meantime, I am hacking away on pepper_34 and pepper_35 support...

**Comment by danielrh** [comment]
Thank you so much for your fast responses! I'm so glad this project is well supported: in this thread you've convinced me to use it for my next open source project

**Comment by danielrh** [comment]
The pepper_35 experiment hasn't turned out so well: I think the llvm-3.2 version of lli doesn't work as expected somehow---
I'm wondering if the #0 is a line number or some sort of other directive that lli is not ignoring.

This is the full error message I'm getting

```
/home/daniel/Downloads/nacl_sdk/nacl_sdk/pepper_35/toolchain/linux_pnacl/bin/pnacl-lli: /tmp/tmpH2D5nM.ll:733:20: error: expected '{' in function body
define i32 @main() #0 {
                   ^
FAIL: Running the generated program failed!
Traceback (most recent call last):
  File "/home/daniel/dev/pepper_35/emscripten/emcc", line 1690, in <module>
    final = shared.Building.emscripten(final, append_ext=False, extra_args=extra_args)
  File "/home/daniel/dev/pepper_35/emscripten/tools/shared.py", line 1223, in emscripten
    assert os.path.exists(filename + '.o.js') and len(open(filename + '.o.js', 'r').read()) > 0, 'Emscripten failed to generate .js: ' + str(compiler_output)
AssertionError: Emscripten failed to generate .js: 
ERROR    root: bootstrapping relooper failed. You may need to manually create relooper.js by compiling it, see src/relooper/emscripten
```

**Comment by danielrh** [comment]
I tried using the emscripten-fastcomp package (current master) and it all started working with pepper_35! The only oddity is that the random color selection looked different with voronoi
http://graphics.stanford.edu/~danielrh/voronoi.png than with pepper_34 or prior. Of course that makes me a bit nervous: what if the math is broken somehow, but otherwise the pepper_35 build appears to be working using
LLVM=/home/daniel/dev/emscripten-fastcomp/build/Release/bin
and NACL_SDK_ROOT=/home/daniel/Downloads/nacl_sdk/nacl_sdk/pepper_35

**Comment by danielrh** [comment]
pepper_34 also has the same odd color selection (in both debug and release) when I build with emscripten-fastcomp so it's nothing new with pepper_35 and looks correct when I build with the standard toolchain (which doesn't work with pepper_35)

**Comment by danielrh** [comment]
Ok I'm silly: the reason the colors look different is that the old rand_r() function looked like:

```
function _rand_r(seed) { // XXX ignores the seed
      return Math.floor(Math.random()*0x80000000);
}
```

And the new one:

```
function _rand_r(seedp) {
    var val = ((Math_imul(((HEAP32[((seedp)>>2)])|0), 31010991)|0) + 0x676e6177 ) & 2147483647; // assumes RAND_MAX is in bit mask form (power of 2 minus 1)
    HEAP32[((seedp)>>2)]=val;
    return val|0;
}
```

so it generates different visual color effects.
So no bug in any toolchain: just a different implementation of rand

**Comment by ncbray** [comment]
It turns out that pepper_35 updates LLVM to 3.4.  The pinned version of Emscripten won't work with that (I tweaked it to work with 3.3, but now I see where this path is heading).  I suspect the only workable solution is to stop trying to use PNaCl's toolchain - the release schedules are not going to line up and there will always be this tug of war.  Since I started the project, I know that someone has cobbled together a "Windows SDK" for Emscripten that may solve the toolchain problem I was originally trying to address.  If that's workable, I'll change the install instructions to require the user install Emscripten themselves.

Hmmm... I'll look closer at the rand_r function.  Different is OK, but the colors make the worry the result is biased in some way.
